cmd/clusterctl/client/cluster/cert_manager.go

```diff
@@ -89,21 +81,71 @@ type CertManagerClient interface {
 
 // certManagerClient implements CertManagerClient .
 type certManagerClient struct {
-	configClient        config.Client
-	proxy               Proxy
-	pollImmediateWaiter PollImmediateWaiter
+	configClient                       config.Client
+	proxy                              Proxy
+	pollImmediateWaiter                PollImmediateWaiter
+	embeddedCertManagerManifestVersion string
+	embeddedCertManagerManifestHash    string
 }
 
 // Ensure certManagerClient implements the CertManagerClient interface.
 var _ CertManagerClient = &certManagerClient{}
 
+func (cm *certManagerClient) setManifestHash() error {
+	yamlData, err := manifests.Asset(embeddedCertManagerManifestPath)
+	if err != nil {
+		return err
+	}
+	cm.embeddedCertManagerManifestHash = fmt.Sprintf("%x", sha256.Sum256(yamlData))
+	return nil
+}
+
+func (cm *certManagerClient) setManifestVersion() error {
+	// Gets the cert-manager objects from the embedded assets.
+	objs, err := cm.getManifestObjs()
+	if err != nil {
+		return err
+	}
+	found := false
+
+	for i := range objs {
+		o := objs[i]
+		if o.GetKind() == "CustomResourceDefinition" {
+			labels := o.GetLabels()
+			version, ok := labels[certmanagerVersionLabel]
+			if ok {
+				s := strings.Split(version, "-")
+				cm.embeddedCertManagerManifestVersion = s[2]
+				found = true
+				break
+			}
+		}
+	}
+	if !found {
+		return errors.Errorf("Failed to detect cert-manager version by searching for label %s in all CRDs", certmanagerVersionLabel)
+	}
+	return nil
+}
+
 // newCertMangerClient returns a certManagerClient.
 func newCertMangerClient(configClient config.Client, proxy Proxy, pollImmediateWaiter PollImmediateWaiter) *certManagerClient {
-	return &certManagerClient{
+	log := logf.Log
+
+	cm := &certManagerClient{
 		configClient:        configClient,
 		proxy:               proxy,
 		pollImmediateWaiter: pollImmediateWaiter,
 	}
+	err := cm.setManifestVersion()
+	if err != nil {
+		log.Info("failed to set Manifest Version")
```

Comment:
Is this the behavior we want?

In this case we'd be proceeding with an empty `cm.embeddedCertManagerManifestVersion`. @fabriziopandini This may be fine because we'll error out in other places but just wanted to confirm.

At the very least we should log the error out as well. Also log messages should start with capital letter. E.g. `Failed to set manifest version`.
```suggestion
		log.Error(err, "Failed to set Manifest Version")
```

Reply:
ok, will update to Error

Reply:
And, if we calculate the hash/version on the fly
does that mean we actually might compare 2 incorrect thing but because they came from place then wrongly think the test pass?

Reply:
Yeah...I see your point. I guess, it depends on what we want to test. So if the behavior we want to test is that _when  `newCertManagerClient` is called, it should set the `embeddedCertManagerManifestVersion` and `embeddedCertManagerManifestHash`_.

So for this behavior the simplest test we can do is ensure that these fields are not empty.
But we can also compute the actual values rather than storing the hardcoded values in the test.
Because in the end, even the hardcoded values are manually computed in the same way. That is, we ran sha256sum on the asset file manually and manually inspected the file for the version.

Hope that clarifies my point regarding the comment https://github.com/kubernetes-sigs/cluster-api/pull/3707#discussion_r510441882

Reply:
@wfernandes could you please check latest code (I modified only one place) so that I can calculate the hash on the fly and compare it with the value from cm so that it will be not empty and has value ... if it works, then I will modify the other tests ..thanks a lot for the detailed comments and help